JDK-8256844: Make NMT late-initializable

tstuefe opened the pull request:
Short: this patch makes NMT available in custom-launcher scenarios and during gtests. It simplifies NMT initialization. It adds a lot of NMT-specific testing, cleans them up and makes them sideeffect-free.

---------

NMT continues to be an extremely useful tool for SAP to tackle memory problems in the JVM.

However, NMT is of limited use due to the following restrictions:

- NMT cannot be used if the hotspot is embedded into a custom launcher unless the launcher actively cooperates. Just creating and invoking the JVM is not enough, it needs to do some steps prior to loading the hotspot. This limitation is not well known (nor, do I believe, documented). Many products don't do this, e.g., you cannot use NMT with IntelliJ. For us at SAP this problem limits NMT usefulness greatly since our VMs are often embedded into custom launchers and modifying every launcher is impossible.
- Worse, if that custom launcher links the libjvm *statically* there is just no way to activate NMT at all. This is the reason NMT cannot be used in the `gtestlauncher`.
- Related to that is that we cannot pass NMT options via `JAVA_TOOL_OPTIONS` and `-XX:Flags=<file>`.
- The fact that NMT cannot be used in gtests is really a pity since it would allow us to both test NMT itself more rigorously and check for memory leaks while testing other stuff.

The reason for all this is that NMT initialization happens very early, on the first call to `os::malloc()`. And those calls happen already during dynamic C++ initialization - a long time before the VM gets around parsing arguments. So, regular VM argument parsing is too late to parse NMT arguments.

The current solution is to pass NMT arguments via a specially prepared environment variable: `NMT_LEVEL_<PID>=<NMT arguments>`. That environment variable has to be set by the embedding launcher, before it loads the libjvm. Since its name contains the PID, we cannot even set that variable in the shell before starting the launcher.

All that means that every launcher needs to especially parse and process the NMT arguments given at the command line (or via whatever method) and prepare the environment variable. `java` itself does this. This only works before the libjvm.so is loaded, before its dynamic C++ initialization. For that reason, it does not work if the launcher links statically against the hotspot, since in that case C++ initialization of the launcher and hotspot are folded into one phase with no possibility of executing code beforehand.

And since it bypasses argument handling in the VM, it bypasses a number of argument processing ways, e.g., `JAVA_TOOL_OPTIONS`.

------

This patch fixes these shortcomings by making NMT late-initializable: it can now be initialized after normal VM argument parsing, like all other parts of the VM. This greatly simplifies NMT initialization and makes it work automagically for every third party launcher, as well as within our gtests.

The glaring problem with late-initializing NMT is the NMT malloc headers. If we rule out just always having them (unacceptable in terms of memory overhead), there is no safe way to determine, in os::free(), if an allocation came from before or after NMT initialization ran, and therefore what to do with its malloc headers. For a more extensive explanation, please see the comment block `nmtPreInit.hpp` and the discussion with @kimbarrett and @zhengyu123 in the JBS comment section.

The heart of this patch is a new way to track early, pre-NMT-init allocations. These are tracked via a lookup table. This was a suggestion by Kim and it worked out well.

Changes in detail:

- pre-NMT-init handling:
	- the new files `nmtPreInit.hpp/cpp` take case of NMT pre-init handling. They contain a small global lookup table managing C-heap blocks allocated in the pre-NMT-init phase.
	- `os::malloc()/os::realloc()/os::free()` defer to this code before doing anything else.
	- Please see the extensive comment block at the start of `nmtPreinit.hpp` explaining the details.

- Changes to NMT:
	- Before, NMT initialization was spread over two phases, `initialize()` and `late_initialize()`. Those were merged into one and simplified - there is only one initialization now which happens after argument parsing.
	- Minor changes were needed for the `NMT_TrackingLevel` enum - to simplify code, I changed NMT_unknown to be numerically 0. A new comment block in `nmtCommon.hpp` now clearly specifies what's what, including allowed level state transitions.
	- New utility functions to translate tracking level from/to strings added to `NMTUtil`
	- NMT has never been able to handle virtual memory allocations before initialization, which is fine since os::reserve_memory() is not called before VM parses arguments. We now assert that.
	- All code outside the VM handling NMT initialization (eg. libjli) has been removed, as has the code testing it.

- Gtests:
	- Some existing gtests had to be modified: before, they all changed global state (turning NMT on/off) before testing. This is not allowed anymore, to keep NMT simple. Also, this pattern disturbed other tests.
	- The new way to test is to passively check whether NMT has been switched on or off, and do tests accordingly: if on, full tests, if off, test just what makes sense in off-state. That does not disturb neighboring tests, gives us actually better coverage all around.
	- It is now possible to start the gtestlauncher with NMT on! Which additionally gives us good coverage.
	- To actually do gtests with NMT - since it's disabled by default - we now run NMT-enabled gtests as part of the hotspot jtreg NMT wrapper. This pattern we have done for a number of other facitilites, see all the tests in test/hotspot/jtreg/gtest.. . It works very well.
	- Finally, a new gtest has been written to test the NMT preinit lookup map in isolation, placed in `gtest/nmt/test_nmtpreinitmap.cpp`.
	
- jtreg:
	- A new test has been added, `runtime/NMT/NMTInitializationTest.java`, testing NMT initialization in the face of many many VM arguments.

-------------

Tests:
- ran manually all new tests on 64-bit and 32-bit Linux
- GHAs
- The patch has been active in SAPs test systems for a while now.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256844](https://bugs.openjdk.java.net/browse/JDK-8256844): Make NMT late-initializable


### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to caf76522d9f8fa5ec71e78b681f826de9c28be0f
 * [Zhengyu Gu](https://openjdk.java.net/census#zgu) (@zhengyu123 - **Reviewer**) ⚠️ Review applies to 42f2ca2243b85198ab20b0ee86a4c8f094d691d8


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4874/head:pull/4874` \
`$ git checkout pull/4874`

Update a local copy of the PR: \
`$ git checkout pull/4874` \
`$ git pull https://git.openjdk.java.net/jdk pull/4874/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4874`

View PR using the GUI difftool: \
`$ git pr show -t 4874`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4874.diff">https://git.openjdk.java.net/jdk/pull/4874.diff</a>

</details>

tstuefe commented:
> Before looking at this, have you checked the startup performance impact?
> 
> Thanks,
> David
> -----

Hi David,

performance should not be a problem. The potentially costly thing is the underlying hashmap. But we keep it operating with a very small load factor.

More details:

Adding entries is O(1). Since during pre-init phase almost only adds happen, startup time is not affected. Still, to make sure this is true, I did a bunch of tests:

- tested WCT of a helloworld, no differences with and without patch
- tested startup time in various of ways, no differences
- repeated those tests with 25000 (!) VM arguments, the only way to influence the number of pre-init allocations. No differences (VM gets slower with and without patch).

----

The expensive thing is lookup since we potentially need to walk a very full hashmap. Lookup affects post-init more than pre-init.

To get an idea of the cost of a too-full preinit lookup table, I modified the VM to do a configurable number of pre-init test-allocations, with the intent of artificially inflating the lookup table. Then, after NMT initialization, I measured the cost of lookup. The short story, I was not able to measure anything, even with a million pre-init allocations. Of course, with more allocations lookup table got fuller and the VM got slower, but the time increase was caused by the cost of the malloc calls themselves, not the table lookup.

Finally, I did an isolated test for the lookup table, testing pure adding and retrieval cost with artificial values. There, I could see costs for add were static (as expected), and lookup cost increased with table population. On my machine:

| lu table entries            | time per lookup |
| ------ |:-------------:|
| 1000    | 3 ns         | 
| 1 mio   | 240 ns       | 

As you can see, if lookup table population goes beyond 1 mio entries, lookup time starts being noticeable over background noise. But with these numbers, I am not worried. Standard lookup population should be around *300-500*, with very long command lines resulting in table populations of *~1000*. We should never seen 10000 entries, let alone millions of them.

Still, I added a jtreg test to verify the expected hash table population. To catch errors like an unforeseen mass of pre-init allocations (lets say a leak or badly written code sneaked in), or if the hash algorithm suddenly is not good anymore.

Two more points

1) I kept this coding deliberately simple. If we are really worried about a degenerated lookup table, we can do things to fix that:
 - we could automatically resize and rehash
 - we could, if we sense something wrong, just stop filling it and disable NMT, stopping NMT init phase prematurely at the cost of not being able to use NMT. 
 
The latter I had implemented already but removed it again to keep complexity down, and because I saw no need.

2) In our propietary production VM we have a system similar to NMT, but predating it. In that system we don't use malloc headers but store all (millions of) malloc'ed pointers in a big hash map. It performs excellent on *all our libc variants*. It is so fast that we just leave it always switched on. This solution has been productive since >10 years, and therefore I am confident that this is viable. This proposed hashmap with a planned population of 300-1000 is really not much :)

dholmes-ora commented on src/hotspot/share/services/memTracker.cpp:
```diff
@@ -174,6 +144,8 @@ bool MemTracker::transition_to(NMT_TrackingLevel level) {
 void MemTracker::error_report(outputStream* output) {
   if (tracking_level() >= NMT_summary) {
     report(true, output, MemReporterBase::default_scale); // just print summary for error case.
+    output->print("Preinit state:");
```
print_cr? Or need space after ':'  ?

dholmes-ora commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+//     be avoided - should use raw malloc(3).
+//
+// We use a basic open hashmap, dimensioned generously - hash collisions should be very rare.
+//  The table is customized for holding malloced pointers. One main point of this map is that we do
```
Nit: double-space at line start (and for rest of comment block)

dholmes-ora commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+class NMTPreInitAllocationTable {
+
+  // table_size: keep table size a prime and the hash function simple; this
+  //  seems to give a good distribution for malloce'd pointers on all our libc variants.
```
typo: malloce'd

dholmes-ora commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+      NMTPreInitAllocation* a = NMTPreInitAllocation::do_alloc(size);
+      add_to_map(a);
+      (*rc) = a->payload();
+      _num_mallocs_pre ++;
```
style nit: no space with unary operator

dholmes-ora commented on src/java.base/share/native/libjli/java.c:
```diff
@@ -806,8 +806,6 @@ CheckJvmType(int *pargc, char ***argv, jboolean speculative) {
 static void
 SetJvmEnvironment(int argc, char **argv) {
```
This doesn't seem to do anything any more.

tstuefe replied:
Thanks for this hint. This brought me to JDK-8186483, which can probably get rolled back now entirely.

dholmes-ora commented on test/hotspot/gtest/nmt/test_nmtpreinitmap.cpp:
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2017, 2020, Oracle and/or its affiliates. All rights reserved.
```
New file should only have single, current, copyright year

dholmes-ora reviewed (commented):
Hi Thomas,

I had a look through this and it seems reasonable, but I'm not familiar enough with the details to approve at this stage.

A few nits below.

Thanks,
David

coleenp reviewed (approved):
This is an interesting and it seems a better way to solve this problem.   Where were you all those years ago??  I hope @zhengyu123 has a chance to review it.
Also interesting is that we were wondering how we could return malloc'd memory on JVM creation failure, and this might partially help with that larger problem.

kimbarrett commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+
+  // Returns start of the user data area
+  void* payload()             { return this + 1; }
+  const void* payload() const { return this + 1; }
```
[Not a review, just a drive-by comment.]  This is a normal and idiomatic overload on the const-ness of `this`.

tstuefe replied:
To expand on Kim's answer. This is a way to solve const/nonconst problems like https://github.com/openjdk/jdk/pull/4938/files#r679639391. 

You get a const version (suitably returning a write-protected pointer) which the compiler chooses in const code, and a non-const version for non-const code, and no awkward const-casts are needed from the outside.

In this case the implementation is simple enough that I just duplicated it; were it more complex, I'd call one in terms of the other. We do this in other places too, see e.g. `ResourceHashTable::lookup_node`.

coleenp replied:
This is ok. This was just new to me.

kimbarrett commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+
+  NMTPreInitAllocation** find_entry(const void* p) const {
+    const unsigned index = index_for_key(p);
+    NMTPreInitAllocation** aa = (NMTPreInitAllocation**) (&(_entries[index]));
```
[Not a review, just a drive-by comment.]  It's casting away const.  Better would be a const_cast.  And probably moved to the final result, with the body keeping things const-qualified.  And maybe const and non-const overloads of this function.  Or maybe this function shouldn't be const-qualified if a non-const result is always needed, but that doesn't seem likely.

tstuefe replied:
I'll rethink this.

tstuefe commented:
> This is an interesting and it seems a better way to solve this problem. Where were you all those years ago?? I hope @zhengyu123 has a chance to review it.

Thank you! I was here, but we were not yet doing much upstream :) To be fair, this problem got quite involved and did cost me some cycles and false starts. I fully understand that the first solution uses the environment variable approach.

I spend some time investigating different ideas with this one; at first I did not use a hash-table but a static pre-allocated buffer from which I fed early allocations. But the code got too complex, and Kim's suggestion with the side table turned out just to be a lot simpler.

> Also interesting is that we were wondering how we could return malloc'd memory on JVM creation failure, and this might partially help with that larger problem.

Yes, this would be trivial now, to return that memory. Though I am afraid it would be a small part only. But NMT may be instrumental in releasing all memory, since it knows everything. Only, its not always enabled.

Is that a real-life problem? Are there cases where the launcher would want to live on if the JVM failed to load?

tstuefe commented:
> Hi Thomas,
> 
> I had a look through this and it seems reasonable, but I'm not familiar enough with the details to approve at this stage.
> 
> A few nits below.
> 
> Thanks,
> David

I did not expect a quick review for this one, so thanks for looking at this! All your suggestions make sense, I'll incorporate them.

..Thomas

zhengyu123 commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+        void* p_new = do_os_malloc(new_size);
+        ::memcpy(p_new, a->payload(), MIN2(a->size, new_size));
+        (*rc) = p_new;
+        _num_reallocs_pre_to_post++;
```
post-NMT-init counter updates are racy.

tstuefe replied:
True, this is racy. It's just for diagnostics though - I rather remove them than make them atomic since we would pay for this with every malloc. Or, maybe atomic + debug only?

zhengyu123 replied:
Either one is fine.

zhengyu123 reviewed (commented):
Sorry for late review. 

Did a quick scan and have a few questions, will do more detail reading later.

tstuefe commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+    NMTPreInitAllocation** aa = find_entry(p);
+    assert((*aa) != NULL, "Entry not found: " PTR_FORMAT, p2i(p));
+    NMTPreInitAllocation* a = (*aa);
+    (*aa) = (*aa)->next;         // remove from its list
```
The code is implicitly thread-safe because the hashmap is only modified in the pre-NMT-init phase. After NMT initialization, the table is read-only. During pre-NMT-init we are effectively single-threaded - at most two threads access the map, the thread loading the libjvm, and the thread calling CreateJavaVM, but not at the same time.

See also the asserts in the AllStatic class `NMTPreInit`.

(I should have described it more clearly, will add a comment.)

zhengyu123 replied:
So, you are saying that there is no memory that is malloc'd pre-NMT-init phase and freed post-NMT-init phase?

zhengyu123 replied:
Okay, I see. It just leaks those memory, so the table can be read-only.

tstuefe replied:
Exaclty. There are a few allocs which either get free'd or realloc'ed post-init, but not enough to make freeing them worth it if it means having to serialize access to the lookup table.

tstuefe commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+// For the lookup table, design considerations are:
+//   - lookup speed is paramount since lookup is done for every os::free() call.
+//   - insert/delete speed only matters for VM startup - after NMT initialization the lookup
+//     table is readonly
```
The point is *after NMT initialization* - `find_and_remove` only gets called before NMT initialization; after that, we only do non-modifying lookup. You'll find the logic in `NMTPreInit::handle_realloc()` and `NMTPreInit::handle_free()`, respectively.

The basic idea behind this is that we remove pointers from the map as long as we can without risking concurrency issues, which is until NMT initialization. After that, we leave the map alone. It was either that or protect the map with a lock, and this is the lesser of two evils since the map is usually sparsely populated.

tstuefe commented:
> Sorry for late review.
> 
> Did a quick scan and have a few questions, will do more detail reading later.

Thanks a lot, I appreciate your feedback!

zhengyu123 commented on src/hotspot/share/services/nmtPreInit.hpp:
```diff
@@ -0,0 +1,353 @@
+  const index_t invalid_index = -1;
+
+  static unsigned calculate_hash(const void* p) {
+    uintptr_t tmp = p2i(p);
```
malloc memory usually is 2-machine word aligned, maybe tmp = tmp >> LP64_ONLY(4) NOT_LP64(3) can result better hash distribution?

tstuefe replied:
I thought so too at first, but it is not necessary. The hash function uses all input bits:
```
    uintptr_t tmp = p2i(p);
    unsigned hash = (unsigned)tmp
                     LP64_ONLY( ^ (unsigned)(tmp >> 32));
```
p2i is lossless since input and result have the same size.
Then,
- for 32-bit, it does not matter since unsigned is the same size as uintptr_t and we lose no bits
- for 64-bit, we xor the upper half of the 64-bit input value into the lower half; again, all input bits count toward the result. We don't gain anything by shifting, we would just exchange the lower 2 bits always being zero against the upper 2 bits always being zero.

---

BTW, I experimented with different hash functions, e.g. http://www.concentric.net/~Ttwang/tech/inthash.htm, but did not really get a better distribution. This somewhat mirrors my experiences when I tried to optimize your hash function in the MallocSiteTable. Seems malloc return value is already "random enough". I coupled it with a crooked table size though, made it a prime.

zhengyu123 reviewed (approved):
Looks good in general.

tstuefe commented:
Hi Reviewers,

thanks for all the feedback!

New version:
- Found an embarrassing bug where I had forgotten to actually ::free preinit allocations. I went through the motions and all but forgot the final call. This would not have been a big problem, only a few scraps of memory lost, but still. Fixed with https://github.com/openjdk/jdk/pull/4874/commits/e2d40f2fbb6ee09c41c2036d37ce9bf0873ed213
- There is a piece of test code which I originally thought had to reside inside the hotspot but realized it did not. I removed it from the hotspot proper and moved it wholesale into a new gtest. https://github.com/openjdk/jdk/pull/4874/commits/caf76522d9f8fa5ec71e78b681f826de9c28be0f
- I added another jtreg test to test that JDK tools can be called with -J-XX:NativeMemoryTracking and that NMT works https://github.com/openjdk/jdk/pull/4874/commits/dc340897d1a5c50af4683943484e4f42478cc7e9
- Resolved @dholmes-ora feedback, including removing more obsolete code from libjli https://github.com/openjdk/jdk/pull/4874/commits/48176dcc2afede91df92121027f4710d2638a41b
- Resolved feedback from @coleenp and @kimbarrett , fixing constness issue with `NMTPreInitAllocationTable::find_entry()`. I provided two versions based on const overloading. Changes trickle up the chain, but I think its better now (e.g. const find operation return const pointers) https://github.com/openjdk/jdk/pull/4874/commits/ef24874a5cc30966eecba86ae407ab58ebeeb7a7
- Resolved @zhengyu123 feedback and removed the post-init counters. https://github.com/openjdk/jdk/pull/4874/commits/ae56cb4d0b679452e8846205f4e43c1ccf4a727e

Thanks for your review work! 

..Thomas

coleenp commented:
> Is that a real-life problem? Are there cases where the launcher would want to live on if the JVM failed to load?

There are a lot of other reasons why the launcher couldn't live on if the JVM fails to load.  This was only one of them.  We used to think about this problem once but don't really think it's solveable.

coleenp commented on src/hotspot/share/services/nmtPreInit.cpp:
```diff
@@ -166,55 +166,12 @@ void* NMTPreInit::do_os_malloc(size_t size) {
-};
-static TestAllocations g_test_allocations; // make this an automatic object to let ctor run during in C++ dynamic initialization.
-#endif // ASSERT
-
```
Oh, good.  I was hoping this didn't need to be an in-jvm test.

coleenp reviewed (approved):
This looks good.  Thanks for fixing the mysterious (to me) cast.

tstuefe commented:
> This looks good. Thanks for fixing the mysterious (to me) cast.

Thank you, Coleen!

tstuefe commented:
Nightlies at SAP showed no problems for several runs now. The failed GHA test (StringDeduplication) seems to have nothing to do with my patch.

@zhengyu123 are you fine with the latest version of this patch?

zhengyu123 commented:
> Nightlies at SAP showed no problems for several runs now. The failed GHA test (StringDeduplication) seems to have nothing to do with my patch.
> 
> @zhengyu123 are you fine with the latest version of this patch?

Still good.

tstuefe commented:
Thanks @coleenp and @zhengyu123 !

/integrate
